Gentoo/Funtoo portage support

trizt opened the pull request:
Converts some of the portage files into osquery tables.
Shows correct version when querying os_version.
Makes it also possible to build on Gentoo/Funtoo

osqueryer commented:
Can one of the osquery admins reply with "ok to test" to kick off a build...

osqueryer commented:
Can one of the osquery admins reply with "ok to test" to kick off a build...

osqueryer commented:
Can one of the osquery admins reply with "ok to test" to kick off a build...

PoppySeedPlehzr commented:
ok to test

osqueryer commented:
:no_good: The commit a0c84020c01f143e2044dddd474b0581901e91f5 (Job results: [512](https://jenkins.osquery.io/job/osqueryPullRequestAudit/512/)) failed the code audit.

theopolis commented:
A few notes on the design.

I think all of the portage-related code would work well in a single `./osquery/tables/system/linux/portage_packages.cpp` implementation file. If you need small helper functions you can inline/static them easily.

What do you think about merging a lot of the table concepts into a powerful `portage_packages` table? In this case `masked` becomes a column with values `1` or `0` (treated as a boolean), the same with `unmasked`. The `keyword` to `package, version` relation becomes singular, and you space-delimit all into a `keywords` column.

theopolis commented:
To fix the code auditing, use `make format_master`

trizt commented:
Thanks for the suggestions. I'll be fixing things during the weekend, got just really confused by the jenkins error. 

A question, should I make a second commit or squash everything into one commit as before?

theopolis commented:
@trizt you can submit multiple commits :) we'll squash them in the Github UI when ready to merge.

trizt commented:
@theopolis trying to get my head around the super table (merge of keywords, mask, unmask), as at the moment I have picked all the lines in the files, without trying to merge configuration lines and I don't see any clear path in merging those, as I see that I may need to make multiple rows as the versions may not match at all in the three files and each can include more than once each package.

For example you could say you want to mask all versions of gnome2 before 3.10 and all versions above version 3.20, which would result two rows in the packages.mask file. I'll try to think more about it and see if I find a way that won't make too many extra rows in a such mega table.

osqueryer commented:
:no_good: The commit 3353529946c6f7724c26b6cc128bee784162bee6 (Job results: [542](https://jenkins.osquery.io/job/osqueryPullRequestAudit/542/)) failed the code audit.

osqueryer commented:
:-1: The commit 3353529946c6f7724c26b6cc128bee784162bee6 (Job results: [566](https://jenkins.osquery.io/job/osqueryPullRequestBuild/566/)) failed one or more tests (**Windows**).

trizt commented:
merged portage_mask, portage_unmask into portage_keywords and merged portage_world into portage_packages.

Seemed to fail to build on windows after merging facebook/master

theopolis commented on osquery/tables/system/linux/os_version.cpp:
```diff
@@ -83,17 +86,26 @@ QueryData genOSRelease() {
 QueryData genOSVersion(QueryContext& context) {
-  if (isReadable(kOSRelease)) {
+  // Funtoo has an empty os-release file
+  if (isReadable(kOSRelease) && boost::filesystem::file_size(kOSRelease)) {
```
Ouch, this is an odd scenario.

For this method, I'd recommend something like:
```cpp
QueryData genOSVersion(QueryContext& context) {
  if (isReadable(kOSRelease)) {
    boost::system::error_code ec;
    // Funtoo has an empty os-release file.
    if (boost::filesystem::file_size(kOSRelease, ec) > 0) {
      return genOSRelease();
    }
  }

  Row r;
  std::string content;
  if (readFile(kRedhatRelease, content).ok()) {
    r["platform_like"] = "rhel";
  } else if (readFile(kGentooRelease, content).ok()) {
    r["platform_like"] = "gentoo";
  } else {
    return {};
  }

  boost::algorithm::trim_all(content);

  [...]
```

theopolis commented:
Sorry it's taken me a while to get back to this PR with comments.

I like the condensing work you've done a lot! This looks much better!

What do you think about further reducing `use_on` and `use_off` into `portage_packages`? I'm also struggling with `portage_make`'s usefulness. As a good rule of thumb, we try to avoid `key`/`value` type tables. In those cases we find osquery isn't adding much value vs. a really simple parse. Going forward we'll have a much more powerful `/etc/` configuration parser with `augeas` too. This doesn't mean we have to remove `package_make`, but let me know if you have strong feelings towards it's inclusion.

theopolis commented on osquery/tables/system/linux/portage.cpp:
```diff
@@ -0,0 +1,416 @@
+
+#include "osquery/core/conversions.h"
+
+#define fs boost::filesystem
```
Prefer:
```cpp
namespace fs = boost::filesystem;
```

theopolis commented on osquery/tables/system/linux/portage.cpp:
```diff
@@ -0,0 +1,416 @@
+namespace tables {
+
+/* File paths used by portage */
+fs::path portagePackageDir = "/var/db/pkg";
```
Let's use `const std::string`s for these configuration options.

For global-constant objects we prefix the lower-camel-case name with a `k`. So this becomes `kPortagePackageDir`. Also, consider using uniform initialization syntax:
```cpp
const std::string kPortagePackageDir{"/var/db/pkg"};
```

theopolis commented on osquery/tables/system/linux/portage.cpp:
```diff
@@ -0,0 +1,416 @@
+/**
+ * @brief container to hold data before we store it to QueryData
+ */
+struct portagePackageHolder {
```
This needs some work, for classes and structs use upper-camel-case: `PortagePackage`, and we can avoid "Holder" since the first part is pretty clear. :)

For these non-POD structs please inherit from `boost::non_copyable` to help prevent copying all of the complex objects.

theopolis commented on osquery/tables/system/linux/portage.cpp:
```diff
@@ -0,0 +1,416 @@
+ * a pair of package name (first) and package version (second).
+ */
+std::pair<std::string, std::string> portageSplitPackageVersion(
+    std::string pkgStr) {
```
Please pass by `const std::string&` for clarity. If you intend to copy the data and modify it, do so in the method implementation.

theopolis replied:
For the Linux/POSIX code, prefer to use snake-case for variable names that aren't global constant. (And NB for class member objects append a `_` to keep track of scope.)

theopolis commented on osquery/tables/system/linux/portage.cpp:
```diff
@@ -0,0 +1,416 @@
+                               const std::string& worldContent) {
+  QueryData results;
+
+  fs::path full_path(fs::initial_path());
```
This is initialized to an unused value.

theopolis commented on osquery/tables/system/linux/portage.cpp:
```diff
@@ -0,0 +1,416 @@
+  QueryData results;
+
+  fs::path full_path(fs::initial_path());
+  full_path = fs::system_complete(fs::path(pkgpath));
```
`pkgpath` was already an `fs::path`, this will incur an unneeded intermediate value initialization.

theopolis replied:
Be very careful with `boost::filesystem` API calls! The simple variant of each will throw exceptions! Either `try` and handle the exception for each or use the `boost::system::error_code` variant.

theopolis commented on osquery/tables/system/linux/portage.cpp:
```diff
@@ -0,0 +1,416 @@
+  fs::path full_path(fs::initial_path());
+  full_path = fs::system_complete(fs::path(pkgpath));
+
+  if (!fs::exists(full_path)) {
```
Same as above, be careful here, you're forgetting to handle an exception.

theopolis commented on osquery/tables/system/linux/portage.cpp:
```diff
@@ -0,0 +1,416 @@
+
+  // store world in a map so we can use find to see if
+  // a package is in the world file
+  std::unordered_map<std::string, bool> world = {};
```
The initialization is unneeded.

theopolis commented on osquery/tables/system/linux/portage.cpp:
```diff
@@ -0,0 +1,416 @@
+  // store world in a map so we can use find to see if
+  // a package is in the world file
+  std::unordered_map<std::string, bool> world = {};
+  if (worldContent.length() > 0) {
```
Maybe `!empty()`

theopolis commented on osquery/tables/system/linux/portage.cpp:
```diff
@@ -0,0 +1,416 @@
+      if (line.size() == 0 || boost::starts_with(line[0], "#")) {
+        continue;
+      }
+      std::pair<std::string, std::string> package =
```
`auto` would make this line pretty. :)

theopolis commented on osquery/tables/system/linux/portage.cpp:
```diff
@@ -0,0 +1,416 @@
+
+  // This is a bit slow, would benefit if it had been multithreaded here
+  fs::directory_iterator end_iter;
+  for (fs::directory_iterator dir_itr(full_path); dir_itr != end_iter;
```
Again, be careful! Maybe use our globbing APIs in `./osquery/filesystem.h`?

trizt replied:
A note for myself:
I tried to access getGlobb() but not accessible as you are supposed to use resolveFilePattern()

theopolis commented on osquery/tables/system/linux/portage.cpp:
```diff
@@ -0,0 +1,416 @@
+  fs::directory_iterator end_iter;
+  for (fs::directory_iterator dir_itr(full_path); dir_itr != end_iter;
+       ++dir_itr) {
+    std::string category = dir_itr->path().c_str();
```
An `fs::path` has a `.string()` method to skip the unneeded initialization.

theopolis commented on osquery/tables/system/linux/portage.cpp:
```diff
@@ -0,0 +1,416 @@
+       ++dir_itr) {
+    std::string category = dir_itr->path().c_str();
+    category.erase(0, 12);
+    if (boost::starts_with(category, "."))
```
Use `{}` even around single-statement blocks.

theopolis commented on osquery/tables/system/linux/portage.cpp:
```diff
@@ -0,0 +1,416 @@
+    category.erase(0, 12);
+    if (boost::starts_with(category, "."))
+      continue;
+    fs::path catpath(fs::initial_path());
```
Unneeded.

theopolis commented on osquery/tables/system/linux/portage.cpp:
```diff
@@ -0,0 +1,416 @@
+    fs::path catpath(fs::initial_path());
+    catpath = fs::system_complete(fs::path(dir_itr->path()));
+
+    if (!fs::exists(catpath) || !isDirectory(catpath)) {
```
Do these need to be `system_complete`?

theopolis commented on osquery/tables/system/linux/portage.cpp:
```diff
@@ -0,0 +1,416 @@
+    for (fs::directory_iterator c_dir_itr(dir_itr->path());
+         c_dir_itr != end_iter;
+         ++c_dir_itr) {
+      Row r;
```
This logic is a bit deep, consider abstracting into another method.

theopolis commented on osquery/tables/system/linux/portage.cpp:
```diff
@@ -0,0 +1,416 @@
+  std::string keywords;
+  std::string masked;
+  std::string unmasked;
+  bool keywordsResult = readFile(portageKeywords, keywords).ok();
```
Instead of creating intermediate booleans, you could have the conditional check `!.empty()` for each.

trizt commented:
@theopolis portage_make, idea was more to make it simple to compare configurations between machines, as this affects how packages are built. I can agree it's not the best table.

I can drop it from this merge and look if I could make something better of it in the future.

Been a bit thinking of that use_on/off, it has it's use, but may not be as good as I thought first (wanted to be simple), really been thinking of a use table (I know you may not be as keen on that), so that you could get package to use, could this way also get the USE flags from the portage_make into one table.

Not decided anything yet, will think a bit about it and make a new pull request with explanation why I did what I did and of course do the fixes you pointed out.

theopolis commented:
You rock @trizt, thanks for taking the time to flush out the design!

trizt commented:
@theopolis 
Changed the portage_use table to only list enabled options which packages has been built with (instead as previously what's configured in /etc/portag/package.use). Split up the use flags to be just one per row, which makes it easier to search for enabled options as you don't need to make a string search ( like = "%useflag%" ). Knowing which use flags was used during build time can help figure out possible vulnerabilities.
Dropped the portage_make (key value table) and see if we at a later time can make something to visualize configuration for Gentoo/Funtoo

theopolis commented:
This looks good @trizt!

My last criticism is mostly on myself, it's a bit difficult for me to follow the logic in `portageSplitPackageVersion`. Mind if I ask one more favor? Can you add a few test cases for that method in `./osquery/tables/system/linux/tests/`, perhaps as `portage_tests.cpp`?

trizt commented:
@theopolis sure I can do that.

osqueryer commented:
:no_good: The commit 97532c7f45b7b3d20580c5cdc1c845c8c41b9ff0 (Job results: [737](https://jenkins.osquery.io/job/osqueryPullRequestBuild/737/)) failed the code audit or documentation test.

osqueryer commented:
:-1: The commit 97532c7f45b7b3d20580c5cdc1c845c8c41b9ff0 (Job results: [3754](https://jenkins.osquery.io/job/osqueryPullRequestBuild/3754/)) failed one or more tests (**Linux**).

theopolis commented on osquery/tables/system/linux/tests/portage_tests.cpp:
```diff
@@ -0,0 +1,64 @@
+
+TEST_F(PortageTests, portageSplitPackageVersion_greater_or_equal_version) {
+
+  auto result = portageSplitPackageVersion("<=sys-kernel/gentoo-sources-4.9.0_alpha2");
```
Guessing this `<=` is a copypasta typo. ;)

trizt replied:
Thanks @theopolis I don't know what I did, I thought I had test built all just moments before, but I must have done some last minute change before I committed.

theopolis commented:
You got it! And sorry for the bad links in the bot messages (fixing that now), but here are the audit results: https://jenkins.osquery.io/job/osqueryPullRequestAudit/742/TargetSystem=aws-ubuntu-16.04/console there are some style nits needed.

trizt commented:
@theopolis nice with the diff in the  console output.

theopolis commented on osquery/tables/system/linux/portage.cpp:
```diff
@@ -0,0 +1,326 @@
+    }
+  }
+
+  for (auto& directory : pkgPaths) {
```
Can this be `const auto&`?

theopolis commented on osquery/tables/system/linux/portage.cpp:
```diff
@@ -0,0 +1,326 @@
+                                            const std::string& masked,
+                                            const std::string& unmasked) {
+  QueryData results;
+  std::unordered_map<std::string, PortagePackage*> tempStorage;
```
This is scary, we shouldn't be using raw pointers.

trizt replied:
@theopolis 
This was to avoid the copy so we could use boost::non_copyable for the PortagePackage.
If you have a suggestion on how to do this without copy and raw pointers, I would happy to change it again.

trizt replied:
@theopolis will shared_ptr be ok?

trizt replied:
@theopolis thanks, that was a lot more than I had expected. Never used anything else than raw pointers, which made me a bit unsure.

theopolis commented on osquery/tables/system/linux/portage.cpp:
```diff
@@ -0,0 +1,326 @@
+      if (line.size() == 0 || boost::starts_with(line[0], "#")) {
+        continue;
+      }
+      PortagePackage* r = new PortagePackage();
```
Either use a `shared_ptr<PortagePackage>` or use `std::move` when emplacing, but let's avoid raw pointer risk.

theopolis commented on osquery/tables/system/linux/portage.cpp:
```diff
@@ -0,0 +1,326 @@
+    std::string pkg = directory;
+    pkg.erase(0, 12);
+
+    std::pair<std::string, std::string> package =
```
nit, can you `auto package` to fit on a single line?

theopolis commented on osquery/tables/system/linux/portage.cpp:
```diff
@@ -0,0 +1,326 @@
+        continue;
+      }
+      PortagePackage* r = new PortagePackage();
+      std::pair<std::string, std::string> package =
```
nit, `auto package =`.

theopolis commented on osquery/tables/system/linux/portage.cpp:
```diff
@@ -0,0 +1,326 @@
+  QueryData results;
+  std::unordered_map<std::string, PortagePackage*> tempStorage;
+
+  if (keywords.length() > 0) {
```
nit, `if (!keywords.empty()) {`

theopolis commented on osquery/tables/system/linux/portage.cpp:
```diff
@@ -0,0 +1,326 @@
+    }
+  }
+
+  if (masked.length() > 0) {
```
nit `if (!masked.empty()) {`

theopolis commented on osquery/tables/system/linux/portage.cpp:
```diff
@@ -0,0 +1,326 @@
+        continue;
+      }
+
+      std::unordered_map<std::string, PortagePackage*>::iterator it =
```
nit, `auto it = `.

Though, we usually use:
```cpp
if (tempStorage.count(line[0]) > 0) {
  tempStorage[line[0]] = "1";
```

theopolis commented on osquery/tables/system/linux/portage.cpp:
```diff
@@ -0,0 +1,326 @@
+        it->second->mask = "1";
+      } else {
+        PortagePackage* r = new PortagePackage();
+        std::pair<std::string, std::string> package =
```
nit, `auto package =`

theopolis commented on osquery/tables/system/linux/portage.cpp:
```diff
@@ -0,0 +1,326 @@
+    }
+  }
+
+  if (unmasked.length() > 0) {
```
nit, `if (!unmasked.empty()) {`.

theopolis commented on osquery/tables/system/linux/portage.cpp:
```diff
@@ -0,0 +1,326 @@
+        continue;
+      }
+
+      std::unordered_map<std::string, PortagePackage*>::iterator it =
```
Same nit as the above iterator usage.

theopolis commented on osquery/tables/system/linux/portage.cpp:
```diff
@@ -0,0 +1,326 @@
+        it->second->unmask = "1";
+      } else {
+        PortagePackage* r = new PortagePackage();
+        std::pair<std::string, std::string> package =
```
ditto.

theopolis commented on osquery/tables/system/linux/portage.cpp:
```diff
@@ -0,0 +1,326 @@
+    }
+  }
+
+  for (auto it = tempStorage.begin(); it != tempStorage.end(); ++it) {
```
If using a `shared_ptr` or move semantics for the second item, this can take the normal range-based iterator loop form.

theopolis commented on osquery/tables/system/linux/portage.cpp:
```diff
@@ -0,0 +1,326 @@
+QueryData parsePortageUseContent(const std::vector<std::string>& pkgPaths) {
+  QueryData results;
+
+  for (auto& directory : pkgPaths) {
```
nit, `const auto&`.

theopolis commented on osquery/tables/system/linux/portage.cpp:
```diff
@@ -0,0 +1,326 @@
+    std::string pkg = directory;
+    pkg.erase(0, 12);
+
+    std::pair<std::string, std::string> package =
```
nit, `auto package`.

theopolis commented:
ok to test
